Better gesture typing accept words whose starting char is close to first gesture path corner

doodhout opened the pull request:
I tried improving the gesture typing mechanics by not requiring the user to start their swiped word with the correct first character.

Example:
- Before this change: you want to swipe the word Juggernaut, but you start the gesture on the key next to it. Result: you will never get the word you wanted, because AnySoftKeyboard dismisses all words not starting with the first letter of the gesture path.
- After this change: AnySoftKeyboard now also suggests words whose path aligns with the gesture path without requiring the first letter to be correct, so when you swipe the gesture path for Juggernaut but you start on the letter K, now you still get the suggestion for the word Juggernaut, since the gesture path in total aligns best with that word.

Commit:
Count a coordinate on the right or bottom edge of a key as still inside the key, because otherwise that exact coordinate would fall in-between keys and thus would not result in a valid gesture path.

Commit:
Modify GestureTypingDetector to also consider candidate words whose first character isn't necessarily the same as the character of the key on which the gesture path began, to not require from the gesture typer to always start the gesture path on the correct key, but allow for slight mistakes.

Functionality:
- only consider candidate words whose starting character is at most the value of START_KEY_PROXIMITY_THRESHOLD_SQUARED away from the gesture's first corner
- add extra penalty distance for candidate words whose first character is not equal to the start key of the gesture path

Tests:
- modify existing tests to allow for candidate words whose first character isn't equal to the gesture's path starting character
- add new tests to cover new functionality

menny commented on ime/app/src/main/java/com/anysoftkeyboard/gesturetyping/GestureTypingDetector.java:
```diff
@@ -26,6 +26,19 @@ public class GestureTypingDetector {
+   * This allows for imprecise gesture starts while filtering obviously wrong candidates.
+   * Value is approximately 1.5 key widths squared (assuming ~100 pixel keys).
+   */
+  private static final int START_KEY_PROXIMITY_THRESHOLD_SQUARED = 22500;
```
This should be a resource. Probably add that to ime/app/src/main/res/values/dimens.xml. For example, instead of 100 pixels, you may add `<dimen name="gesture_typing_proximity_threshold">20dp</dimen>` (or whatever value works for you). Then you can use `context.getResources().getDimensionPixelSize(R.dimen.gesture_typing_proximity_threshold)`

Or something like that.

menny replied:
150 dp depends on the device. Mine will be 450 pixels.
<img width="524" height="316" alt="Screenshot 2025-12-15 at 3 37 17 PM" src="https://github.com/user-attachments/assets/69d1fe38-2a14-4d52-ba36-2dc8d7884265" />

Consider experimenting with different values to see how performance and accuracy changes.

menny reviewed (changes requested):
Thank you so much! This is very helpful. Can you attend the comments?

doodhout commented on ime/app/src/main/java/com/anysoftkeyboard/keyboards/Keyboard.java:
```diff
@@ -612,9 +612,9 @@ public boolean isInside(int x, int y) {
       final boolean bottomEdge = (edgeFlags & EDGE_BOTTOM) != 0;
       return (x >= this.x || (leftEdge && x <= this.x + this.width))
-          && (x < this.x + this.width || (rightEdge && x >= this.x))
+          && (x <= this.x + this.width || (rightEdge && x >= this.x))
```
Good catch, I realise now that I tried fixing the 1 pixel horizontal gap from the test case in completely the wrong way.

Original reasoning: one of the new unit-tests picks a spot in the middle of two keys, but of course that spot falls into the gap between the keys, causing the GestureTypeDetector's check `if (k.isInside(corners[0], corners[1])) {` to not match any key, thus `startKey` is null, thus it will return zero candidates.

I will have to revert this change and rethink how to handle that situation, because if it's acceptable to not start on the correct key, then it should also be acceptable to not start on any key.

Commit:
- remove requirement that gesture should start on a key, because a gesture could also start in-between keys when there is sufficient space in between

Commit:
- reverting mistake where I (wrongly) compensated for the 1-pixel horizontal gap between keys by considering key-bounds 1 pixel larger
- exaggerated key and row gap sizes in unit-test to make this behaviour more obvious

Commit:
- moved static final int START_KEY_PROXIMITY_THRESHOLD_SQUARED to dimens.xml and passed that value through via the constructor. (but still need to fix the unit-tests)

doodhout commented:
I'm working through your suggestions.
I need to fix some existing unit-test and then need to add these requested unit-tests.
Thanks for your patience. :-)

Commit:
Revert "- remove requirement that gesture should start on a key, because a gesture could also start in-between keys when there is sufficient space in between"

This reverts commit 4a1f4067daae0027e5ecb8c35e200ab17913012b.

Commit:
- remove requirement that gesture should start on a key, because a gesture could also start in-between keys when there is sufficient space in between

Commit:
- increased start key proximity threshold from 20 pixels to 150 pixels
- fix failing unit-test

Commit:
- created new unit-tests

Commit:
- added change in dimens.xml from 20dp to 150dp, because I forgot that in the previous commit

Commit:
- removing new unit-tests since this functionality is already thoroughly covered in the existing added unit-tests
- fix existing unit-tests by increasing MAX_SUGGESTIONS to allow for candidates with a different starting letter

Commit:
- ran bazel run //:format to auto-format the changes

menny commented on ime/app/src/main/java/com/anysoftkeyboard/gesturetyping/GestureTypingDetector.java:
```diff
@@ -318,21 +333,23 @@ public ArrayList<String> getCandidates() {
+        }
+
+        // Calculate squared distance from gesture start to word's starting key
+        final int distanceSquared = wordStartKey.squaredDistanceFrom(corners[0], corners[1]);
```
Micro-optimization:
first check if the `wordStartKey != startKey` and if so then calculate the distance and do the other stuff.

menny replied:
If you are not worried about this, or do not agree, lmk

menny reviewed (changes requested):
this is great, I have a few questions

Commit:
- only if wordStartKey is not equal to startKey, then calculate distance of word start key from gesture first corner

doodhout commented:
> @doodhout I will continue your work in #4624 Thank you so much for this effort!

Hi @menny, sorry for the radio silence, last month has been very intense for me.

If you require assistance, I should be able to provide this during the coming weekend.

Thanks for the feedback and the patience!
